server/src/controllers/christmas-tree/permits.es6

```diff
@@ -165,26 +142,24 @@ at pay.gov ${updatedPermit.paygovError}`
 const recordPayGovError = (error, result, res, permit, requestType) => {
   logger.error(`ERROR: ServerError: Pay.gov- ${error}. Updating permit with error while ${requestType}`);
   return paygov.getResponseError(requestType, result)
-    .then(paygovError => updatePermitWithError(res, permit, paygovError));
+    .then(paygovError => updatePermit(permit, {
+      status: 'Error',
+      paygovError: JSON.stringify(paygovError)
+    })
+      .then(updatedPermit => res.status(400).json(formatPermitError(updatedPermit))));
 };
 
 /**
  * @function grabAndProcessPaygovToken - Private function to error objects from the permit's errors
  * @param {Object} xmlDatafromPayGov - xml from pay.gov
  * @param {Object} permit - permit object from database
  * @param {Object} res - http response
- * @return {Object} - promise of whether permit was updated with token
+ * @return {Promise} - resolves if permit was updated with token
  */
-const grabAndProcessPaygovToken = (payGovXmlRes, permit, res) => new Promise((resolve, reject) => {
-  xml2jsParse(payGovXmlRes, (parseErr, result) => {
-    if (!parseErr) {
-      paygov.getToken(result)
-        .then(token => resolve(updatePermitWithToken(res, permit, token)))
-        .catch(error => recordPayGovError(error, result, res, permit, 'startOnlineCollection'));
-    }
-    reject(parseErr);
-  });
-});
+const grabAndProcessPaygovToken = (payGovXmlRes, permit, res) => util.parseXml(payGovXmlRes)
+  .then(result => paygov.getToken(result)
+    .then(token => updatePermitWithToken(res, permit, token))
+    .catch(error => recordPayGovError(error, result, res, permit, 'startOnlineCollection')));
```

Comment:
This leaves the caller responsible for handling an error with the xml parsing. In this case the caller DOES handle the error but does not record the paygov error. Because of where the `catch` is we're only recording this error when it occurs as a result of attempting to update the permit in the database.

Reply:
That makes sense but I think you want the error handling to be more explicit, or at least add a clear comment. My first impression reading this code was: "oh, the indenting is off". *fix fix fix* "oh, the catch is in the wrong place" *fix fix fix* and suddenly I've changed the behavior by accident. So it's the subtlety of it that causes bad behavior.

Do you have tests for this so that if someone was to accidentally change it, it would be caught?

Reply:
I feel you there. There are tests for some of those conditions, I'm not sure that all of them are handled. I will add a comment for now that addresses what appears to be the intent, i don't have a good sense of whether the intent is actually the right approach or not.